javascript/ql/lib/semmle/javascript/frameworks/data/internal/SharedModelValidation.qll

```diff
@@ -0,0 +1,143 @@
+        "regex-use%",
+        // JavaScript-only currently, but may be shared in the future
+        "credentials-%",
+        // Swift-only currently, but may be shared in the future
+        "%string-%length", "weak-hash-input-%"
+      ])
+  }
+}
+
+/** An outdated models-as-data sink kind. */
+class OutdatedSinkKind extends string {
+  OutdatedSinkKind() {
+    this =
+      [
+        "sql", "url-redirect", "xpath", "ssti", "logging", "groovy", "jexl", "mvel", "xslt", "ldap",
+        "pending-intent-sent", "intent-start", "set-hostname-verifier", "header-splitting", "xss",
+        "write-file", "create-file", "read-file", "open-url", "jdbc-url", "command-line-injection",
+        "code", "html", "remote", "uncontrolled-format-string", "js-eval"
+      ]
+  }
+
+  /** Gets a replacement kind for an outdated sink kind. */
+  private string replacementKind() {
+    this = ["sql", "xpath", "groovy", "jexl", "mvel", "xslt", "ldap", "code", "html"] and
+    result = this + "-injection"
+    or
+    this = "js-eval" and result = "code-injection"
+    or
+    this = "url-redirect" and result = "url-redirection"
+    or
+    this = "ssti" and result = "template-injection"
+    or
+    this = "logging" and result = "log-injection"
+    or
+    this = "pending-intent-sent" and result = "pending-intents"
+    or
+    this = "intent-start" and result = "intent-redirection"
+    or
+    this = "set-hostname-verifier" and result = "hostname-verification"
+    or
+    this = "header-splitting" and result = "response-splitting"
+    or
+    this = "xss" and result = "html-injection\" or \"js-injection"
+    or
+    this = ["write-file", "remote"] and result = "file-content-store"
+    or
+    this = ["create-file", "read-file"] and result = "path-injection"
+    or
+    this = ["open-url", "jdbc-url"] and result = "request-forgery"
+    or
+    this = "command-line-injection" and result = "command-injection"
+    or
+    this = "uncontrolled-format-string" and result = "format-string"
+  }
+
+  /** Gets an error message for an outdated sink kind. */
+  string outdatedMessage() {
+    result =
+      "The kind \"" + this + "\" is outdated. Use \"" + this.replacementKind() + "\" instead."
+  }
+}
+
+/** A valid models-as-data source kind. */
+class ValidSourceKind extends string {
+  bindingset[this]
+  ValidSourceKind() {
+    this =
+      [
+        // shared
+        "local", "remote",
+        // Java
+        "android-external-storage-dir", "contentprovider",
+        // C#
+        "file", "file-write",
+        // JavaScript
+        "database-access-result", "remote-flow"
+      ]
+    or
+    this.matches([
+        // shared
+        "qltest%", "test-%",
+        // Swift
+        "%string-%length"
+      ])
+  }
+}
+
+/** A valid models-as-data summary kind. */
+class ValidSummaryKind extends string {
+  ValidSummaryKind() {
+    this =
+      [
+        // shared
+        "taint", "value",
+        // JavaScript
```

Comment:
```suggestion
        // Dynamic languages
```
Support for this is shared between the dynamic languages.

Reply:
Updated in https://github.com/github/codeql/pull/13324/commits/3bfb5f9ac4696ee208947e6344061a007386dbb0.